File: kotlin/factorymethod.kt
abstract class Animal {
    abstract fun speak(): String
}

class Dog : Animal() {
    override fun speak() = "Woof!"
}

class Cat : Animal() {
    override fun speak() = "Meow!"
}

object AnimalFactory {
    fun createAnimal(type: String): Animal? = when (type) {
        "dog" -> Dog()
        "cat" -> Cat()
        else -> null
    }
}

File: assignment_4/csharp/factorymethod.cs
using System;

abstract class Animal {
    public abstract string Speak();
}

class Dog : Animal {
    public override string Speak() => "Woof!";
}

class Cat : Animal {
    public override string Speak() => "Meow!";
}

class AnimalFactory {
    public static Animal? CreateAnimal(string type) {
        return type switch {
            "dog" => new Dog(),
            "cat" => new Cat(),
            _ => null
        };
    }
}
